### PR TITLE
Replace NULL with C++11 nullptr

### PR DESCRIPTION
**Pull request by leekillough**
Replace most uses of `NULL` with the C++11 `nullptr`.

In C++11 and later, `nullptr` is encouraged instead of `NULL` or `0`, to represent null pointers.

`nullptr` is a C++11 keyword (no header needed) which represents a _null pointer constant_. It can be implicitly converted to any pointer type, and can be compared to any pointer type.

For the rationale of `nullptr`, see N2431, _[A name for the null pointer: nullptr](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2431.pdf)_. One reason, is that the integer constant `0` can present ambiguity when passed to a function -- if a function can accept both `int` and pointer, then if `0` or C++ `NULL` is passed, the `int` overload is selected, while if `nullptr` is passed, the pointer overload is selected.

In C++, `NULL` is defined as `0`, unlike in C, where `NULL` is defined as `((void*)0)`, because in C++, `((void*)0)` is not _implicitly_ convertible to any other pointer type, while the null pointer constant `0` or `nullptr` is.

This PR brings the Chapel source code in line with C++11 `nullptr` usage standards.

All C++ uses of `NULL` have been changed to `nullptr`.

`NULL`, when used in the Chapel language itself, in compiler diagnostic messages, and in generated `C` code, remains `NULL`. Only in the C++11 source code, is `NULL` changed to `nullptr`.

There are a few obvious cases of comparison of pointers to `0` which have been replaced with comparison to `nullptr`, or converted to a Boolean test such as `if (!ptr)`.

~~`NULL` in the Bison parser code (`chapel.ypp`) has been replaced with the `YY_NULLPTR` macro, which is defined by Bison in a language-aware way (C/C++/C++11).~~

### CODE REVIEW

**Comment by bradcray** [comment]
Paraphrasing earlier comments from Slack:

I don’t feel strongly about this change either way and am willing to go with whatever the group decides (I understand the change reflects a best practice).

If this is something we want to adhere to over time, we should also add checks to prevent people from injecting new `NULL`s into the code, at least for our common compilers.  

(Doing the obvious Google search:  For example, could we add `-Wzero-as-null-pointer-constant` to our set of `g++` and `clang++` developer flags?  And have the GitHub checks test this build if they don't already?)

File: compiler/parser/Makefile [hunk 1]
```diff
@@ -71,15 +71,16 @@ flex_version_check:
 #
 include $(COMPILER_ROOT)/make/Makefile.compiler.foot
 
+NO_NULLPTR_ERROR=-Wno-error=zero-as-null-pointer-constant
```
Comment (bradcray) [inline compiler/parser/Makefile:74]: If I'm reading this right, it seems that it's being set regardless of the compiler being used, where not all compilers / compiler versions may recognize it.  For that reason, It seems like it should be set in the various compiler-specific `Makefile.*` files.

File: make/compiler/Makefile.clang [hunk 2]
```diff
@@ -123,10 +123,11 @@ endif
 #
 # Flags for turning on warnings for C++/C code
 #
-WARN_CXXFLAGS = -Wall -Werror -Wpointer-arith -Wwrite-strings -Wno-strict-aliasing -Wmissing-declarations
+WARN_CXXFLAGS = -Wall -Werror -Wpointer-arith -Wwrite-strings -Wno-strict-aliasing -Wmissing-declarations -Werror=zero-as-null-pointer-constant
 WARN_CFLAGS = $(WARN_CXXFLAGS) -Wmissing-prototypes -Wstrict-prototypes -Wmissing-format-attribute
 WARN_GEN_CFLAGS = $(WARN_CFLAGS)
 SQUASH_WARN_GEN_CFLAGS = -Wno-unused -Wno-uninitialized
+NO_NULLPTR_ERRROR = -Wno-error=zero-as-null-pointer-constant
```
Comment (bradcray) [inline make/compiler/Makefile.clang:130]: ERRROR -> ERROR  (and in Makefile.gnu) as well

File: make/compiler/Makefile.gnu [hunk 3]
```diff
@@ -162,10 +162,11 @@ GEN_CFLAGS += $(C_STD)
 # of Clang header files.
 #
 WARN_COMMONFLAGS = -Wall -Werror -Wpointer-arith -Wwrite-strings -Wno-strict-aliasing
-WARN_CXXFLAGS = $(WARN_COMMONFLAGS) -Wno-comment
+WARN_CXXFLAGS = $(WARN_COMMONFLAGS) -Wno-comment -Werror=zero-as-null-pointer-constant
```
Comment (bradcray) [inline make/compiler/Makefile.gnu:165]: Will this flag be supported by all versions of g++ that we test against, or do we need to restrict it to certain versions?

**Comment by leekillough** [comment]
> If this is something we want to adhere to over time, we should also add checks to prevent people from injecting new `NULL`s into the code, at least for our common compilers.
> 
> (Doing the obvious Google search: For example, could we add `-Wzero-as-null-pointer-constant` to our set of `g++` and `clang++` developer flags? And have the GitHub checks test this build if they don't already?)

That flag was added yesterday, before I read your comment, and the changes were committed this morning.

File: make/compiler/Makefile.clang (hunk 2 above)
Comment (leekillough) [reply make/compiler/Makefile.clang:130]: My terminal must have been messed up (often a bad terminal state causes characters to be repeated without showing them on the screen).

File: compiler/parser/Makefile (hunk 1 above)
Comment (leekillough) [reply compiler/parser/Makefile:74]: Yes, it belongs in the compiler-specific `Makefile`s`.

File: make/compiler/Makefile.gnu (hunk 3 above)
Comment (leekillough) [reply make/compiler/Makefile.gnu:165]: > Will this flag be supported by all versions of g++ that we test against, or do we need to restrict it to certain versions?

That's a good question.  GCC added in version 4.7. Clang has it in version 5.0. Do we support older versions?

Comment (bradcray) [reply make/compiler/Makefile.gnu:165]: I'm not the best person to ask about that (and would take it to the chapel slack channel).  I don't believe we use gcc < 4.7.  I'm less certain about clang.

**Comment by leekillough** [comment]
With the latest changes, the lexical analyzer output created by Flex is edited by `sed` to replace `NULL` with `nullptr`, to avoid seeing spurious warnings.

Even though the output of Flex is reentrant C code, we name the file `flex-chapel.cpp` and compile it as C++, and so the function names are mangled as in C++, and we get the `nullptr` warnings, although the lexer is not compiled with `-Werror`.

The `sed` Regex syntax used was chosen to be compatible with both BSD/MacOS and Linux `sed`. MacOS `sed` does not support `\b` for word boundaries, but has some other word boundary extensions which are not supported on Linux. I choose the POSIX common denominator.

If you look at the diffs, you'll see some cases where they aren't simply `NULL` -> `nullptr` or `0` -> `nullptr` substitutions, for example:

* When a source file can be compiled as both C and C++, `#ifdef __cplusplus` is used to differentiate them and choose either `NULL` or `nullptr`.
* When a templated class (such as in `map.h`) uses keys/values/etc. which can either be pointers or integers, the C++11 syntax `C{}` is used to create a zero instance of that type, rather than having to test types and choose between `0`, `nullptr` or `NULL`. `C{}` creates a 0-initialized instance of type `C`.
* In places where class pointer members are initialized as null pointers in constructors, the C++11 member initialization syntax `ptrType member{};`or `ptrType member = nullptr;` is used, and if there is no other initialization needed in the default constructor, the default constructor is written as `className() = default;` This makes the code more compact and clear.

All `paratest` tests pass. If `NULL` or `0` is used instead of `nullptr` in the Chapel compiler or runtimes, an error will be raised.

Although tests like `if (ptr != nullptr)` are more self-documenting, tests like simply `if (ptr)` or `if (!ptr)` are okay, and do not raise any warnings with respect to this PR.

**Comment by leekillough** [comment]
Closed due to lack of interest.

**Pull request closed without merging**